### ISSUE TITLE
Pytorch build fails on Ubuntu 14.04

### ISSUE THREAD

**Issue by skckompella** [pytorch/pytorch#2093]
OS: Ubuntu 14.04
Python version: 3.6
Cuda version: v8
CuDNN version: v6
CMAKE version: 3.6

When I try to build Pytorch, I get the following error and compilation fails:

> torch/csrc/cuda/Module.cpp:9:18: fatal error: nccl.h: No such file or directory
>  #include <nccl.h>

More of the log can be seen in the attached file
[log.txt](https://github.com/pytorch/pytorch/files/1147066/log.txt)

Please let me know where I am going wrong.

**Comment by ngimel** [comment]
Apparently, you have system-wide install of nccl, but you don't have nccl.h in accessible location. How did you install nccl? If you installed runtime deb package, you'd also need to install dev package (https://github.com/NVIDIA/nccl/releases/download/v1.3.4-1/libnccl-dev_1.3.4-1.cuda8.0_amd64.deb)

**Comment by skckompella** [comment]
Thank you for the response. I installed nccl as below:
```
git clone https://github.com/NVIDIA/nccl.git
cd nccl/
make CUDA_HOME=/usr/local/cuda

sudo make install
sudo mkdir -p /usr/local/include/external/nccl_archive/src
sudo ln -s /usr/local/include/nccl.h /usr/local/include/external/nccl_archive/src/nccl.h
```

I tried to install the .deb file you pointed to but got the error:

> Selecting previously unselected package libnccl-dev.
> (Reading database ... 288164 files and directories currently installed.)
> Preparing to unpack libnccl-dev_1.3.4-1.cuda8.0_amd64.deb ...
> Unpacking libnccl-dev (1.3.4-1+cuda8.0) ...
> dpkg: dependency problems prevent configuration of libnccl-dev:
>  libnccl-dev depends on libnccl1 (= 1.3.4-1+cuda8.0); however:
>   Version of libnccl1 on system is 1.2.3-1+cuda8.0.
> 
> dpkg: error processing package libnccl-dev (--install):
>  dependency problems - leaving unconfigured
> Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
> Errors were encountered while processing:
>  libnccl-dev

Can you please let me know how to upgrade libnccl1?

**Comment by ngimel** [comment]
Just uninstall whatever you installed, pytorch comes with its own copy of nccl.

**Comment by skckompella** [comment]
I was getting the build error before the install as well. Hence, I tried to install it. I will uninstall nccl and paste the build log here.

**Comment by skckompella** [comment]
My bad.. cleaning up everything actually worked! Thanks again! 
 closing...

**Comment by Eniac-Xie** [comment]
Hi, @skckompella , I just have the same error, what do you mean by "cleaning up everything",remove /usr/local/include/nccl.h and nccl libs under /usr/local/lib? Thank you

**Comment by Eniac-Xie** [comment]
I remove the files above, then compile pytorch, another error occurs:

[ 69%] Built target gloo
[ 73%] Building NVCC (Device) object gloo/CMakeFiles/gloo_cuda.dir/nccl/gloo_cuda_generated_nccl.cu.o
In file included from /home/xiecw/pytorch/torch/lib/gloo/gloo/nccl/nccl.cu:10:0:
/home/xiecw/pytorch/torch/lib/tmp_install/include/gloo/nccl/nccl.h:12:18: fatal error: nccl.h: No such file or directory
 #include <nccl.h>
                  ^
compilation terminated.

**Comment by soumith** [comment]
@Eniac-Xie  can you put your full log file of install in https://gist.github.com/ and give us a link to it.

**Comment by Eniac-Xie** [comment]
Hi @soumith , thanks for your response.

I found the reason of the error I said above, but I do not know how to handle it.

I used release v0.2.0 code for installing, if I use master branch and remove the nccl file I said above, I install pytorch successfully.
But I want to use v0.2.0 for compatibility purpose, is there some way to solve this preblem? I found that pytorch/torch/lib/gloo/third-party/nccl is empty in v0.2.0 code, can I just copy the https://github.com/pytorch/pytorch/tree/master/torch/lib/nccl and compile?
